image metadata property

montegoulding opened the pull request.

Commit:
Merge branch 'develop' into feature/image_metadata

Commit:
[[ ImageMetadata ]] Implemented read only image metadata array property

Commit:
[[ Image Metadata ]] Added docs, renamed metadata_of_field, implemented rounding on density

livecodeian commented on engine/src/image.cpp:
```diff
@@ -1115,6 +1127,8 @@ Exec_stat MCImage::setprop(uint4 parid, Properties p, MCExecPoint &ep, Boolean e
 			char *t_name = nil;
 			IO_handle t_stream = nil;
+            MCImageMetadata t_metadata;
+            MCMemoryClear(&t_metadata, sizeof(t_metadata));
```
Hi Monte, this change seems to be unnecessary as t_metadata isn't used here.

montegoulding replied:
Sorry about that. Was an early change I forgot to cleanup after working out how things worked...

livecodeian commented on engine/src/imageloader.cpp:
```diff
@@ -96,6 +96,19 @@ bool MCImageLoader::GetFrameCount(uint32_t &r_frame_count)
+	return true;
+    
+}
+
```
Note that this will replace the contents of r_metadata with the contents of m_metadata, so you should ensure that m_metadata is cleared in the MCImageLoader::MCImageLoader() constructor, otherwise it may contain uninitialised data.

montegoulding replied:
Ah, OK, thanks

Commit:
[[ ImageMetadata ]] Removed an unnecessary code and added initialisation of m_metadata

trevordevore commented:
@runrevian What is the status of this pull request? Any chance of getting it incorporated into the engine? Currently on OS X we can't determine the proper dimensions to display retina images at when importing 3rd party image files into LiveCode.

livecodeian commented on engine/src/image.cpp:
```diff
@@ -916,6 +917,17 @@ Exec_stat MCImage::getprop(uint4 parid, Properties which, MCExecPoint& ep, Boole
+        if (m_rep != nil)
+        {
+            MCImageMetadata t_metadata;
+            (static_cast<MCLoadableImageRep *>(m_rep))->GetMetadata(t_metadata);
```
This cast is dangerous, as the underlying type of m_rep can easily be something other than MCLoadableImageRep

montegoulding replied:
Ah, forgot to go back and change that. It should be cached rep.

livecodeian commented:
Hi Trevor, there is an outstanding issue with this pull request, and it will also require updating to be mergeable into the current development branch. Paging @montegoulding :)

montegoulding commented:
OK, I'll look at the merge conflicts today but I did resolve and push changes requested back in October. @runrevian am I merging into develop, develop6.7 or develop7.0? It looks like develop and develop6.7 are fairly close now.

Commit:
[[ ImageMetadata ]] Refactor of image metadata property and resolved incorrect cast

Merge commit '1693a9d39bf3728534555a5ca865bead831cbef0' into feature/image_metadata

Conflicts:
	engine/src/ibmp.cpp
	engine/src/igif.cpp
	engine/src/iimport.cpp
	engine/src/ijpg.cpp
	engine/src/image.h
	engine/src/image_rep.h
	engine/src/image_rep_densitymapped.cpp
	engine/src/image_rep_encoded.cpp
	engine/src/imageloader.cpp
	engine/src/imageloader.h
	engine/src/ipng.cpp

montegoulding commented:
@runrevian This is my first play with the refactored property handling. It's simpler once you get your head around it but you may want to be extra diligent checking that I've done everything I needed to do given that if there's any docs I don't know where they are.

livecodeali commented:
The property side of this looks good to me - you managed to pick the most complicated property type to do @montegoulding, especially as you had to add a new macro ;-) 

I'll let @runrevian complete the review from the image object point of view.

livecodeian commented on engine/src/image.cpp:
```diff
@@ -951,6 +953,17 @@ Exec_stat MCImage::getprop_legacy(uint4 parid, Properties which, MCExecPoint& ep
+        {
+            MCImageMetadata t_metadata;
+            (static_cast<MCCachedImageRep *>(m_rep))->GetMetadata(t_metadata);
+            MCImageGetMetadata(ep, t_metadata);
```
This doesn't cover the case of MCMutableImageRep, which is not a cached image rep. Why not just make GetMetaData a pure virtual method of MCImageRep avoiding the need to cast altogether? The MCMutableImageRep implementation can return false, which you would need to check for and handle here.

montegoulding replied:
OK, I'll take another stab... was sure I had all those classes covered but I guess not.

montegoulding replied:
Ah I see, I missed it because it's in a different file.

Commit:
[[ ImageMetadata ]] Moved GetMetadata to a virtual method of MCImageRep

Commit:
Merge commit 'bc7b202c4c6ee7e842cf7338143bcaa83f8d7aab' into feature/image_metadata

Conflicts:
	ide

montegoulding commented:
Thanks @livecodeali it was interesting trying to nut it out... I re-merged because for some reason I ended up committing a dirty ide in the last merge but it seemed to be only an untracked log file which you might want to add to gitignore or add ignore = untracked to gitmodules.

peter-b commented:
@montegoulding: We've ignored/removed that log file now. See:
- runrev/livecode-ide#11
- runrev/livecode-ide#27

montegoulding commented:
Guys there's now merge conflicts on this again which it would be nice if GitHub notified me of but it doesn't apparently. Can you give me a date when this is scheduled to be reviewed so I can ensure there's no conflicts on it?

livecodeian commented:
Hi Monte, I've merged this branch and resolved the remaining conflicts. Thanks.
